### PR TITLE
[test rom] move RESET_INFO to reset_reason

### LINKED ISSUE #13265: Update the test ROM to handle reset reasons the same way as the mask ROM. Write the reset reason to the retention SRAM, then clear the RESET_INFO register.
As @a-will  suggests in #13096, this should get our tests more consistent and compatible with back-door loading or bootstrapping. It may end up touching all the tests that reboot or check reset info, but we're committed to the cleanup so it seems worthwhile and should fix some of our tests that are having trouble with reboot on the cw310

### PR DESCRIPTION
**Pull request by drewmacrae**
- Updates the testplan to reflect the change
- rewrites the reset_info to the reset_reason in the retention sram as mapped in the silicon_creator/lib/driver
- unlike mask_rom doesn't spend testing time to initialize the retention sram on POR or other conditions
- Transitioned other tests to use retention SRAM to identify cause of resets using a testutil
- The retention sram test doesn't check the reset_reason from the silicon_creator/lib/driver map
- drivers are faked for english_breakfast's test-rom (will always appear as if reset was POR)
- sram_ctrl_sleep_sram_ret_contents_test is no longer broken

Fixes #13265

Signed-off-by: Drew Macrae <drewmacrae@google.com>

### CODE REVIEW

**Comment by drewmacrae** [comment]
I'm planning to also make the change in this PR to move the reset_info into reset_reason and fix the affected tests

**Comment by drewmacrae** [comment]
Verilated tests look good, I still need to check and debug the CW310 tests before I think this is ready to merge but I'll take review now if anyone can take a look

File: sw/device/silicon_creator/lib/drivers/rstmgr_functest.c
Comment (alphan) [inline sw/device/silicon_creator/lib/drivers/rstmgr_functest.c:30]: this function does not have a parameter:
```suggestion
  rstmgr_testutils_reason_clear();
```

File: sw/device/lib/testing/rstmgr_testutils.h
Comment (alphan) [inline sw/device/lib/testing/rstmgr_testutils.h:106]:
```suggestion
void rstmgr_testutils_reason_clear(void);
```

File: sw/device/lib/testing/rstmgr_testutils.h
Comment (alphan) [inline sw/device/lib/testing/rstmgr_testutils.h:101]:
```suggestion
dif_rstmgr_reset_info_bitfield_t rstmgr_testutils_reason_get(void);
```

File: sw/device/silicon_creator/lib/drivers/rstmgr_functest.c
Comment (drewmacrae) [reply sw/device/silicon_creator/lib/drivers/rstmgr_functest.c:30]: Indeed

**Comment by drewmacrae** [comment]
I'll need to make the 
sram_ctrl_sleep_sram_ret_contents_test.c
skip checking this is retained across a reboot since after a boot it's updated to reflect the reason and not the data written. Is there anyone specifically I should talk to about missing that bit of test coverage? We still use the ram, we just can't really test it across a reboot with a test that runs in flash.

**Comment by drewmacrae** [comment]
I'm not sure if I can make the retention sram driver compatible with English Breakfast. That's my next hurdle to deal with after I fix the retention SRAM test to not fail when this modifies retention sram on boot.

**Comment by weicaiyang** [comment]
> I'll need to make the sram_ctrl_sleep_sram_ret_contents_test.c skip checking this is retained across a reboot since after a boot it's updated to reflect the reason and not the data written. Is there anyone specifically I should talk to about missing that bit of test coverage? We still use the ram, we just can't really test it across a reboot with a test that runs in flash.

This change LGTM

**Comment by drewmacrae** [comment]
> I'm not sure if I can make the retention sram driver compatible with English Breakfast. That's my next hurdle to deal with after I fix the retention SRAM test to not fail when this modifies retention sram on boot.

I'm going to avoid the OTP read for the test-rom and use a constant mask instead to keep the test_rom compatible with English Breakfast (which appears not to have an OTP controller)

**Comment by drewmacrae** [comment]
I'd like to do something similar to https://github.com/lowRISC/opentitan/pull/12875 here and we should eventually clean it up with the resoltion to #12905

**Comment by drewmacrae** [comment]
I think in the test_rom we should avoid SRAM init all together to shorten tests.

I had some trouble making the rstmgr dif work. It appears to depend on uninitialized structures and calls to it hang during the test_rom's execution. For EB I'll fake the drivers to pass CI.

File: sw/device/lib/testing/test_rom/test_rom.c
Comment (timothytrippel) [inline sw/device/lib/testing/test_rom/test_rom.c:86]: Do you need this comment? It is sort of confusing. Maybe you just want to expand on the comment below:
"Store the reset reason in retention SRAM and clear the register to match mask ROM behavior."

Comment (drewmacrae) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: That's misleading based on old behavior and I should clean it up.

I think it's important that we skip sram init in the test_rom, but the rest of that isn't helpful.

Comment (timothytrippel) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: yep thats right, we should try to skip the init if we can for performance, and it should be ok here since you do a write before any read. If it were the other way around I would expect you to get a bus integrity error (like what happened in #11854) on sim_dv platform (and maybe others too) since the retention SRAM is initialized to garbage (with incorrect integrity bits) in the testbench: https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/dv/env/seq_lib/chip_sw_base_vseq.sv;drc=fb839e641e76466d238b38af4be3d58c5ec9cfb7;l=70.

File: sw/device/lib/testing/test_rom/test_rom.c
Comment (timothytrippel) [inline sw/device/lib/testing/test_rom/test_rom.c:86]: any reason you use the mask rom driver code here instead of the DIF `dif_rstmgr_reset_info_get()`? This driver code seems like it also collects the crashdump info into a global (https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/rstmgr.c;drc=94cdceaa94c19ce1876c03ef7d4c19f69c887c0a;l=73) which might be uncessary?

**Review by timothytrippel — APPROVED** [review]

File: sw/device/lib/testing/test_rom/test_rom.c
Comment (timothytrippel) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: hmm thats odd. Did you make sure to call `dif_rstmgr_init()` to initialize the handle before calling `dif_rstmgr_reset_info_get()`?

Comment (drewmacrae) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: I did not... I'll try that and check all the tests still pass, and then update the EB fakes to cover the functions in the DIF as well.

Comment (drewmacrae) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: It works but will need some more work/indirection to fake the rstmgr DIF for EB.

Comment (drewmacrae) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: #13642 Shows my current thinking on what needs to be done to fake the relevant functions in the rstmgr DIF

Comment (timothytrippel) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: since this is handled in a separate PR, can I merge this one?

Comment (drewmacrae) [reply sw/device/lib/testing/test_rom/test_rom.c:86]: Yeah, that should be good. Thanks!

**Comment by tjaychen** [comment]
does this warrant a wider announcement? especially for any ongoing test writers.

**Pull request merged**
